byoeb-v1/byoeb/byoeb/background_jobs/did_you_know/send_dyk.py

```
@@ -0,0 +1,344 @@
+    if os.getenv("APP_ENV") == "PROD":
+        match_stage["User.user_type"] = "asha"
+    else:
+        match_stage["User.test_user"] = True
+
+    user_client = await user_db_service._get_collection_client(user_db_service.collection_name)
+    dyk_client = await user_db_service._get_collection_client(dyks_sent_collection_name)
+    for user in await user_client.afetch_all(match_stage):
+        uid = user["User"]["user_id"]
+        dyk_records = await dyk_client.afetch_all({"user_id": uid})
+        if dyk_records and any(r.get("status") == "pending" for r in dyk_records):
+            continue
+        sent_dyk_ids = set([r["dyk_id"] for r in dyk_records]) if dyk_records else set()
+        yield (User(**user["User"]), sent_dyk_ids)
+
+
+async def queue(records: Dict[LanguageCode, Dict[str, str]]) -> Tuple[int, int]:
+    """
+    Takes a structured set of DYK records and randomly distributes them across the userbase, ensuring:
+    - a user is sent DYK message in only their language
+    - a user is not sent a DYK message they were previously sent
+    Returns lists of queued and exhausted operations.
+
+    Example:
+        records = {
+            LanguageCode.ENGLISH: {
+                "00000000-0000-0000-0000-000000000000": "Blueberries were once red."
+            }
+        }
+
+        dispatched, exhausted = await dispatch(records)
+        for phone_number, dyk_uuid in dispatched:
+            print("Assigned DYK %s to %s", (dyk_uuid, phone_number))
+
+        for phone_number in exhausted:
+            print("Cannot assign DYK to %s (they received all DYKs)" % phone_number)
+    """
+
+    n_queued = 0
+    n_exhausted = 0
+    lang_sets = {lang: set(messages.keys()) for lang, messages in records.items()}
+    queued_client_ops = []
+    async for user, sent in pick_candidates(records.keys()):
+        lang = LanguageCode(user.user_language)
+        diff = lang_sets[lang] - sent  # deduplication
```

What if this is a Marathi user? We should handle cases where user is onboarded on langX but we do not have dyks for langX as is the current setup for Marathi (unsure about Kannada but that is next).

---

`pick_candidates` excludes users with unsupported languages:
```py
match_stage = {"User.user_language": {"$in": [x.value for x in langs]}}
```
Marathi users will be left out. Looking back at this, logging aggregate stats may be useful here to know the missed DYK audience.

---

About unexpected languages, language resolution happens earlier than the `lang = LanguageCode(user.user_language)` here. E.g., registering a language=xx in bot_config.json causes the script to fail early:
```sh
> poetry run python3 byoeb/background_jobs/did_you_know/send_dyk.py
pydantic_core._pydantic_core.ValidationError: 1 validation error for LangEntry
language
  Input should be 'hi', 'en', 'mr' or 'te' [type=enum, input_value='xx', input_type=str]
    For further information visit https://errors.pydantic.dev/2.12/v/enum
```

Similarly, registering 'mr' when there is no 'mr' in dyk.csv fails early as well:
```sh
ValueError: 'mr' is not in list
```

---

https://github.com/A4i-tech/byoeb/pull/34/commits/c715a97dd946754c0d46ce569b3990795f08828d elaborates the ValueError above:
```diff
- ValueError: 'mr' is not in list
+ ValueError: Column "mr" does not exist in dyk.csv - did you forget to create a column for "mr"?
```